### PR TITLE
Transaction fails if SearchNarrowingInterceptor is registered and Partitioning Enabled - fix cross-tenant requests failure

### PR DESCRIPTION
**Pull request by volodymyr-korzh**
Part of the fix for issue #5388 
Includes fix for cross-tenant request failure

### CODE REVIEW

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java [hunk 1]
```diff
@@ -69,6 +71,55 @@ void massageBaseUrl_givenBaseUrlAndNullTenant_shouldReturnBaseUrl() {
...
+
+		String inputUrl = "TENANT1/Patient/123";
+		String actual = ourTenantStrategy.resolveRelativeUrl(inputUrl, myRequestDetails);
+
+		assertEquals(inputUrl, actual);
+	}
+
+	@Test
+	void resolveRelativeUrl_urlAlreadyContainsDifferentTenantId_returnsInputUrl() {
+		when(myRequestDetails.getTenantId()).thenReturn("TENANT1");
```
Comment (michaelabuckley) [inline hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java:101]: minor: these feel repetitive,  consider a `@ParameterizedTest` with `@CsvSource`
See ca.cdr.api.security.ScopeCalculatorTest for an example.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategy.java [hunk 2]
```diff
@@ -105,9 +105,26 @@ public String massageServerBaseUrl(String theFhirServerBase, RequestDetails theR
...
+			return theRelativeUrl;
+		}
+		String nextToken = tokenizer.peek();
+		// there is no tenant ID in parent request details or tenant ID is already present in URL - skip url resolution
+		if (theRequestDetails.getTenantId() == null || nextToken.equals(theRequestDetails.getTenantId())) {
+			return theRelativeUrl;
+		}
+
+		// token is Resource type - adding tenant ID from parent request details
+		if (isResourceType(nextToken, theRequestDetails)) {
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategy.java:120]: problem: what about operations?

**Review by michaelabuckley — CHANGES_REQUESTED** [review]
Please cover operations too.

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java [hunk 3]
```diff
@@ -69,6 +71,55 @@ void massageBaseUrl_givenBaseUrlAndNullTenant_shouldReturnBaseUrl() {
...
+		when(myFHIRContext.getResourceTypes()).thenReturn(Collections.singleton("Patient"));
+		when(myRequestDetails.getFhirContext()).thenReturn(myFHIRContext);
+
+		String inputUrl = "TENANT2/Patient/123";
+		String actual = ourTenantStrategy.resolveRelativeUrl(inputUrl, myRequestDetails);
+
+		assertEquals(inputUrl, actual);
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java]: Agree - Fixed

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java [hunk 4]
```diff
@@ -69,6 +71,55 @@ void massageBaseUrl_givenBaseUrlAndNullTenant_shouldReturnBaseUrl() {
...
+		when(myRequestDetails.getTenantId()).thenReturn("TENANT1");
+		when(myFHIRContext.getResourceTypes()).thenReturn(Collections.singleton("Patient"));
+		when(myRequestDetails.getFhirContext()).thenReturn(myFHIRContext);
+
+		String inputUrl = "Patient/123";
+		String actual = ourTenantStrategy.resolveRelativeUrl(inputUrl, myRequestDetails);
+
+		assertEquals("TENANT1/" + inputUrl, actual);
+	}
+
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java:94]: Added test case for $export

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java [hunk 5]
```diff
@@ -69,6 +71,55 @@ void massageBaseUrl_givenBaseUrlAndNullTenant_shouldReturnBaseUrl() {
...
+		String inputUrl = "TENANT2/Patient/123";
+		String actual = ourTenantStrategy.resolveRelativeUrl(inputUrl, myRequestDetails);
+
+		assertEquals(inputUrl, actual);
+	}
+
+	@Test
+	void resolveRelativeUrl_urlStartsWithResourceType_tenantIdIsAddedToUrl() {
+		when(myRequestDetails.getTenantId()).thenReturn("TENANT1");
+		when(myFHIRContext.getResourceTypes()).thenReturn(Collections.singleton("Patient"));
```
[the remark this answers is not in the document]
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java]: Using real FhirContext requires adding `hapi-fhir-structures-r4` dependency to hapi-fhir-test-utilities pom. This creates cyclic reference issue during maven build. So I left it as-is.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategy.java (hunk 2 above)
Comment (volodymyr-korzh) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategy.java]: Fixed - now we concatenate tenant for operations as well

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java (hunk 1 above)
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java]: Indeed, with @ParameterizedTest it is easier to read - covered all cases with one Parameterized test

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java [hunk 6]
```diff
@@ -69,6 +74,24 @@ void massageBaseUrl_givenBaseUrlAndNullTenant_shouldReturnBaseUrl() {
 		assertEquals(BASE_URL, actual);
 	}
 
+	@CsvSource(value = {
+			" , ", // empty URL - return input URL
```
Comment (michaelabuckley) [inline hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java:78]: trivial suggestion for future: if you add the comment as a third column to the csv, you can use that message in the assert.  Makes debugging easier.  IntelliJ shows it too in the console.

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java [hunk 7]
```diff
@@ -69,6 +74,24 @@ void massageBaseUrl_givenBaseUrlAndNullTenant_shouldReturnBaseUrl() {
...
+			"TENANT1/Patient/123, TENANT1/Patient/123", // tenant ID already exists - return input URL
+			"TENANT2/Patient/123, TENANT2/Patient/123", // requestDetails contains different tenant ID - return input URL
+			"Patient/123		, TENANT1/Patient/123", // url starts with resource type - add tenant ID to URL
+			"$export			, TENANT1/$export" // url starts with operation - add tenant ID to URL
```
Comment (michaelabuckley) [inline hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java:82]: suggestion:

How about adding:
- TENNANT2/$export
- Patient/$export
- TENNANT1/Patient/$export

I believe your code work fine right _now_, but I'd feel safer against someone else making future changes with a few more tests.

**Review by michaelabuckley — APPROVED** [review]

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java (hunk 6 above)
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java]: fixed, adding one more column indeed makes it more readable, great suggestion

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java (hunk 7 above)
Comment (volodymyr-korzh) [reply hapi-fhir-test-utilities/src/test/java/ca/uhn/fhir/rest/server/tenant/UrlBaseTenantIdentificationStrategyTest.java]: fixed, added these cases as well

**Pull request merged**